File: app/src/main/java/jp/hazuki/yuzubrowser/action/item/TranslatePageSingleAction.java
package jp.hazuki.yuzubrowser.action.item;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Parcel;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Spinner;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.action.SingleAction;
import jp.hazuki.yuzubrowser.action.view.ActionActivity;
import jp.hazuki.yuzubrowser.utils.ArrayUtils;
import jp.hazuki.yuzubrowser.utils.app.StartActivityInfo;

public class TranslatePageSingleAction extends SingleAction implements Parcelable {
    private static final String FIELD_NAME_FROM = "0";
    private static final String FIELD_NAME_TO = "1";
    private String mFrom;
    private String mTo;

    public TranslatePageSingleAction(int id, JsonParser parser) throws IOException {
        super(id);

        if (parser != null) {
            if (parser.nextToken() != JsonToken.START_OBJECT) return;
            while (parser.nextToken() != JsonToken.END_OBJECT) {
                if (parser.getCurrentToken() != JsonToken.FIELD_NAME) return;
                if (FIELD_NAME_FROM.equals(parser.getCurrentName())) {
                    switch (parser.nextToken()) {
                        case VALUE_STRING:
                            mFrom = parser.getText();
                            break;
                        case VALUE_NULL:
                            break;
                    }
                    continue;
                }
                if (FIELD_NAME_TO.equals(parser.getCurrentName())) {
                    switch (parser.nextToken()) {
                        case VALUE_STRING:
                            mTo = parser.getText();
                            break;
                        case VALUE_NULL:
                            break;
                    }
                    continue;
                }
                parser.skipChildren();
            }
        }
    }

    @Override
    public void writeIdAndData(JsonGenerator generator) throws IOException {
        generator.writeNumber(id);
        generator.writeStartObject();
        generator.writeStringField(FIELD_NAME_FROM, mFrom);
        generator.writeStringField(FIELD_NAME_TO, mTo);
        generator.writeEndObject();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeString(mFrom);
        dest.writeString(mTo);
    }

    protected TranslatePageSingleAction(Parcel source) {
        super(source.readInt());
        mFrom = source.readString();
        mTo = source.readString();
    }

    public static final Creator<TranslatePageSingleAction> CREATOR = new Creator<TranslatePageSingleAction>() {
        @Override
        public TranslatePageSingleAction createFromParcel(Parcel source) {
            return new TranslatePageSingleAction(source);
        }

        @Override
        public TranslatePageSingleAction[] newArray(int size) {
            return new TranslatePageSingleAction[size];
        }
    };

    @Override
    public StartActivityInfo showMainPreference(ActionActivity context) {
        return showSubPreference(context);
    }

    @Override
    public StartActivityInfo showSubPreference(ActionActivity context) {
        View view = LayoutInflater.from(context).inflate(R.layout.action_translate_page_setting, null);
        final Spinner fromSpinner = (Spinner) view.findViewById(R.id.fromSpinner);
        final Spinner toSpinner = (Spinner) view.findViewById(R.id.toSpinner);

        String list[] = context.getResources().getStringArray(R.array.translate_language_list);
        final String values[] = context.getResources().getStringArray(R.array.translate_language_values);
        int length = list.length;

        String fromList[] = new String[length + 1];
        fromList[0] = context.getString(R.string.action_translate_auto_select);
        System.arraycopy(list, 0, fromList, 1, length);
        ArrayAdapter<String> from_adapter = new ArrayAdapter<>(context, android.R.layout.simple_spinner_item, fromList);
        from_adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        fromSpinner.setAdapter(from_adapter);

        String toList[] = new String[length + 1];
        toList[0] = context.getString(R.string.action_translate_select_each);
        System.arraycopy(list, 0, toList, 1, length);
        ArrayAdapter<String> to_adapter = new ArrayAdapter<>(context, android.R.layout.simple_spinner_item, toList);
        to_adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        toSpinner.setAdapter(to_adapter);

        int index;
        index = ArrayUtils.findIndexOfValue(mFrom, values) + 1;
        fromSpinner.setSelection(index);
        index = ArrayUtils.findIndexOfValue(mTo, values) + 1;
        toSpinner.setSelection(index);

        new AlertDialog.Builder(context)
                .setTitle(R.string.action_settings)
                .setView(view)
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        int position;
                        position = fromSpinner.getSelectedItemPosition();
                        mFrom = (position == 0) ? "auto" : values[position - 1];
                        position = toSpinner.getSelectedItemPosition();
                        mTo = (position == 0) ? null : values[position - 1];
                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();

        return null;
    }

    public String getTranslateFrom() {
        return mFrom;
    }

    public String getTranslateTo() {
        return mTo;
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/action/item/TranslatePageSingleAction.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.action.item

import android.app.AlertDialog
import android.os.Parcel
import android.os.Parcelable
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Spinner
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.core.utility.utils.ArrayUtils
import jp.hazuki.yuzubrowser.legacy.R
import jp.hazuki.yuzubrowser.legacy.action.SingleAction
import jp.hazuki.yuzubrowser.legacy.action.view.ActionActivity
import jp.hazuki.yuzubrowser.ui.app.StartActivityInfo
import java.io.IOException

class TranslatePageSingleAction : SingleAction, Parcelable {
    var translateFrom: String? = null
    var translateTo: String? = null
        private set

    @Throws(IOException::class)
    constructor(id: Int, reader: JsonReader?) : super(id) {
        if (reader != null) {
            if (reader.peek() != JsonReader.Token.BEGIN_OBJECT) return
            reader.beginObject()
            while (reader.hasNext()) {
                if (reader.peek() != JsonReader.Token.NAME) return
                when (reader.nextName()) {
                    FIELD_NAME_FROM -> {
                        if (reader.peek() == JsonReader.Token.NULL) {
                            reader.nextNull<String>()
                        } else {
                            if (reader.peek() != JsonReader.Token.STRING) return
                            translateFrom = reader.nextString()
                        }
                    }
                    FIELD_NAME_TO -> {
                        if (reader.peek() == JsonReader.Token.NULL) {
                            reader.nextNull<String>()
                        } else {
                            if (reader.peek() != JsonReader.Token.STRING) return
                            translateTo = reader.nextString()
                        }
                    }
                    else -> reader.skipValue()
                }
            }
            reader.endObject()
        }
    }

    @Throws(IOException::class)
    override fun writeIdAndData(writer: JsonWriter) {
        writer.value(id)
        writer.beginObject()
        writer.name(FIELD_NAME_FROM)
        writer.value(translateFrom)
        writer.name(FIELD_NAME_TO)
        writer.value(translateTo)
        writer.endObject()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeString(translateFrom)
        dest.writeString(translateTo)
    }

    private constructor(source: Parcel) : super(source.readInt()) {
        translateFrom = source.readString()
        translateTo = source.readString()
    }

    override fun showMainPreference(context: ActionActivity): StartActivityInfo? {
        return showSubPreference(context)
    }

    override fun showSubPreference(context: ActionActivity): StartActivityInfo? {
        val view = View.inflate(context, R.layout.action_translate_page_setting, null)
        val fromSpinner = view.findViewById<Spinner>(R.id.fromSpinner)
        val toSpinner = view.findViewById<Spinner>(R.id.toSpinner)

        val list = context.resources.getStringArray(R.array.translate_language_list)
        val values = context.resources.getStringArray(R.array.translate_language_values)
        val length = list.size

        val fromList = arrayOfNulls<String>(length + 1)
        fromList[0] = context.getString(R.string.action_translate_auto_select)
        System.arraycopy(list, 0, fromList, 1, length)
        val fromAdapter = ArrayAdapter<String>(context, android.R.layout.simple_spinner_item, fromList)
        fromAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        fromSpinner.adapter = fromAdapter

        val toList = arrayOfNulls<String>(length + 1)
        toList[0] = context.getString(R.string.action_translate_select_each)
        System.arraycopy(list, 0, toList, 1, length)
        val toAdapter = ArrayAdapter<String>(context, android.R.layout.simple_spinner_item, toList)
        toAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        toSpinner.adapter = toAdapter

        var index: Int
        index = ArrayUtils.findIndexOfValue<String>(translateFrom, values) + 1
        fromSpinner.setSelection(index)
        index = ArrayUtils.findIndexOfValue<String>(translateTo, values) + 1
        toSpinner.setSelection(index)

        AlertDialog.Builder(context)
                .setTitle(R.string.action_settings)
                .setView(view)
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    var position = fromSpinner.selectedItemPosition
                    translateFrom = if (position == 0) "auto" else values[position - 1]
                    position = toSpinner.selectedItemPosition
                    translateTo = if (position == 0) null else values[position - 1]
                }
                .setNegativeButton(android.R.string.cancel, null)
                .show()

        return null
    }

    companion object {
        private const val FIELD_NAME_FROM = "0"
        private const val FIELD_NAME_TO = "1"

        @JvmField
        val CREATOR: Parcelable.Creator<TranslatePageSingleAction> = object : Parcelable.Creator<TranslatePageSingleAction> {
            override fun createFromParcel(source: Parcel): TranslatePageSingleAction {
                return TranslatePageSingleAction(source)
            }

            override fun newArray(size: Int): Array<TranslatePageSingleAction?> {
                return arrayOfNulls(size)
            }
        }
    }
}
